pocs/camera/zwo.py

```diff
@@ -0,0 +1,446 @@
+
+        if Camera._ASIDriver is None:
+            # Initialise the driver if it hasn't already been done
+            Camera._ASIDriver = libasi.ASIDriver(library_path=library_path)
+
+        logger = get_root_logger()
+        serial_number = kwargs.get('serial_number')
+        if not serial_number:
+            msg = "Must specify serial_number (string ID) for ZWO ASI Camera"
+            logger.error(msg)
+            raise ValueError(msg)
+
+        logger.debug("Looking for ZWO ASI camera with ID '{}'".format(serial_number))
+
+        if not Camera._ids:
+            # Not cached camera IDs, need to probe for connected cameras
+            n_cameras = Camera._ASIDriver.get_num_of_connected_cameras()
+            if n_cameras == 0:
+                msg = "No ZWO ASI camera devices found"
+                logger.error(msg)
+                warn(msg)
+                return
+
+            # Get the IDs
+            for camera_index in range(n_cameras):
+                # Can get IDs without opening cameras by parsing the name string
+                self._info = Camera._ASIDriver.get_camera_property(camera_index)
+                model, _, id = self._info['name'].partition('(')
+                if not id:
+                    logger.warning("Found ZWO ASI camera with no ID set")
+                    break
+                assert id.endswith(')'), self.logger.error("Expected ID enclosed in ()")
+                id = id[:-1]
+                Camera._ids.append(id)
+
+            logger.debug('Connected ASI ZWO cameras: {}'.format(Camera._ids))
+
+        try:
+            self._camera_index = Camera._ids.index(serial_number)
+        except ValueError:
+            msg = "Could not find ZWO ASI camera with ID '{}'".format(serial_number)
+            logger.error(msg)
+            warn(msg)
+            return
+        else:
+            logger.debug("Found ZWO ASI Camera with ID '{}' at index {}".format(
+                serial_number, self._camera_index))
+
+        if serial_number in Camera._assigned_ids:
+            msg = "ZWO ASI Camera with ID '{}' already in use".format(serial_number)
+            logger.error(msg)
+            warn(msg)
+            return
+
+        super().__init__(name, *args, **kwargs)
+        self.connect()
+        assert self.is_connected, error.PanError("Could not connect to {}".format(self))
+
+        Camera._assigned_ids.append(self.uid)
+
+        self._setter_try('ccd_set_point', 'set_point', set_point)
+
+        if filter_type:
+            # connect() will have set this based on camera info, but that doesn't know about filters
+            # upstream of the CCD. Can be set manually here, or handled by a filterwheel attribute.
+            self._filter_type = filter_type
+
+        self._setter_try('gain', 'gain', gain)
+
+        if image_type:
+            self._setter_try('image_type', 'image_type', image_type)
+        else:
+            # Take monochrome 12 bit raw images by default, if we can
```

Comment:
You say 12 bit and then look for `RAW16`, is that correct?

Reply:
Yes, in a sense. The analogue to digital converter inside the camera produces (at most) 12 bit data, but that gets zero-padded to 16 bits on output to align with conventional 8 bit byte boundaries

It gets even more confusing for the low bit depth option. The raw data format for that is 8 bits per pixel but the camera specs talk about 10 bit mode, so the 2 least significant bits are getting thrown away?